Windows docker agents blog article (#3181)
jenkins-infra/jenkins.io: content/blog/2020/05/2020-05-07-docker-windows-agents.adoc

---
layout: post
title: "Windows Docker Agent Images"
tags:
- announcement
- docker
- platform-sig
author: slide_o_mix
opengraph:
  image: /images/docker/dockerJenkins_social.png
---

We would like to announce the availability of Windows agent images for Docker.

image:/images/docker/dockerJenkins.png[Jenkins and Docker, role=center, float=right, height=224]

== New images

* link:https://hub.docker.com/repository/docker/jenkins/agent[jenkins/agent] is a basic agent which bundles the agent.jar for agent <= => master communication. This is most useful as a base image for other images.

* link:https://hub.docker.com/repository/docker/jenkins/inbound-agent[jenkins/inbound-agent] is an agent that is based on the jenkins/agent image above. It provides a wrapper script written in PowerShell to help specify the parameters to agent.jar.

> 

Reviewer: Another plugins reference on this line

Author: I don't see a plugin reference in this part.

> * link:https://hub.docker.com/repository/docker/jenkins/ssh-agent[jenkins/ssh-agent] is an image which has OpenSSH installed and should be used with the link:plugins.jenkins.io/ssh-slaves/[SSH Build Agents Plugin].

Reviewer: And here

=== jenkins/inbound-agent

The inbound-agent Docker image tries to provide a higher level interaction with the agent.jar executable. It provides a PowerShell wrapper script around agent.jar and it is specified as the entrypoint so that you just need to pass in some command line arguments to run the agent. Several of these can also be passed using environment variables instead (they are listed in all capitals).

Command line arguments include:

* `-Url <URL>` or JENKINS_URL - the Jenkins master URL (this should NOT include the computer/NAME/slave-agent.jnlp part)
* `-Secret <SECRET>` or JENKINS_SECRET - the secret as shown on the master after creating the agent
* `-Name <NAME>` or JENKINS_AGENT_NAME - the name of the agent, it should match the name you specified when creating the agent on the master
* `-Tunnel <HOST:PORT>` or JENKINS_TUNNEL - A tunnel host to route the TCP traffic through when connecting to the master
* `-WorkDir <WORKDIR>` or JENKINS_AGENT_WORKDIR - same as the -workDir parameter mentioned above in the jenkins/agent image information.

> * `-WebSocket` or JENKINS_WEB_SOCKET=true - when present the connection to the master will be done via WebSocket instead of TCP networking.

Reviewer 2: Since WebSocket is a layer over TCP, maybe better to describe the difference in terms of the term that is being used rather than the protocol?
```suggestion
* `-WebSocket` or JENKINS_WEB_SOCKET=true - when present the connection to the master will be done via WebSocket through the Jenkins URL rather than using a separate network port.
```

* `-DirectConnection <HOST:PORT>` or JENKINS_DIRECT_CONNECTION - Connect directly to this TCP agent port, skipping the HTTP(S) connection parameter download.
* `-InstanceIdentity <BASE64VALUE>` or JENKINS_INSTANCE_IDENTITY - The base64 encoded InstanceIdentity byte array of the Jenkins master. When this is set, the agent skips connecting to an HTTP(S) port for connection info

> * `-Protocols <PROTOCOLS>` or JENKINS_PROTOCOLS - Specify the remoting protocols to attempt when instanceIdentity is provided

Reviewer 2: What are the allowed values for protocols?

I assume some way of saying "WebSocket" is one of them.  Is there also some way of saying "Jenkins Remoting"?

Author: Good question, it looks like the Windows agents don't currently actually support that parameter (would be an easy fix to do it). It's an option passed to the agent.jar, I will find out what is supported there and also create a PR to actually use the `-Protocols` value

Author: Added link to the list of protocols in the documentation from the remoting repository and a note about the issue with the Protocols parameter.

* `-JavaHome <JAVA_HOME>` or JAVA_HOME  - An override for the default JAVA_HOME baked into the image (different for JDK8 vs. JDK11)

Example:

```
docker run jenkins/inbound-agent:windowsservercore-1809 -Url http://jenkins-server:port -WorkDir=C:/Users/jenkins/Agent -Secret <SECRET> -Name <AGENTNAME>
```

Example using environment variables:

```
docker run -e "JENKINS_URL=http://jenkins-server:port" -e "JENKINS_AGENT_NAME=AGENTNAME" jenkins/inbound-agent:windowsservercore-1809 -WorkDir=C:/Users/jenkins/Agent -Secret <SECRET> -Name <AGENTNAME>
```

The URL, Name and Secret parameters are required, but can be specified as either command line parameters or environment variables.

=== jenkins/ssh-agent

> As mentioned above the jenkins/ssh-agent docker image is based on SSH communication with the master, rather than the remoting TCP or WebSocket protocols. This communication is initiated by the master rather than the agent connecting to the master is is done with jenkins/agent and jenkins/inbound-agent.

Reviewer 2: Remove a duplicated word
```suggestion
As mentioned above the jenkins/ssh-agent docker image is based on SSH communication with the master, rather than the remoting TCP or WebSocket protocols. This communication is initiated by the master rather than the agent connecting to the master is done with jenkins/agent and jenkins/inbound-agent.
```

The image sets up a `jenkins` user and the OpenSSH server so that the master can connect to the agent via SSH. The image expects an SSH public key as a parameter and puts that key into the authorized_keys file for the jenkins user. The private key should be specified in the agent configuration on the master to allow the master to connect correctly.

Example:

```
docker run jenkins/ssh-agent "<public key>"
```

You can also pass the public key as an environment variable when using `docker run`.

Example:

```
docker run -e "JENKINS_AGENT_SSH_PUBKEY=<public key>" jenkins/ssh-agent
```

> You will then be able to connect this agent using the link:plugins.jenkins.io/ssh-slaves/[SSH Build Agents plugin] as "jenkins" with the matching private key.

Reviewer: Isn’t there a plugin macro that can be used so you don’t need the full url?

Author: Thanks, I didn't know that was there.

---
layout: post

> title: "Windows Docker Agent Images"

Reviewer 3: ```suggestion
title: "Windows Docker Agent Images: general availability"
```

tags:
- announcement
- docker
- platform-sig
author: slide_o_mix
opengraph:
  image: /images/docker/dockerJenkins_social.png
---

We would like to announce the availability of Windows agent images for Docker.

> 

Reviewer 3: ```suggestion
These images allow provisioning Jenkins agents with Windows OS on Docker and Kubernetes.

```

An example of using an image in Jenkins Pipeline would be great., e.g. for inbound-agent below

image:/images/docker/dockerJenkins.png[Jenkins and Docker, role=center, float=right, height=224]

> == New images

Reviewer 4: Yes, I was very confused by this otherwise.

You do not plan to create multiarch images, right?

> == New images

Author: We want to do multiarch eventually, it's on the roadmap

> * link:https://hub.docker.com/r/jenkins/agent[jenkins/agent] is a basic agent which bundles the agent.jar for agent <= => master communication. This is most useful as a base image for other images.

Reviewer 3: ```suggestion
* link:https://hub.docker.com/r/jenkins/agent[jenkins/agent] is a basic agent which bundles the agent.jar for agent <= => master communication. This is most useful as a base image for other images.
  Windows images are available starting from version link:https://github.com/jenkinsci/docker-agent/releases/tag/4.3-4[4.3-4]
```

Reviewer 4: Not sure about Windows usage, but at least the Linux flavor is usable as is via the command launcher.

> * link:https://hub.docker.com/r/jenkins/inbound-agent[jenkins/inbound-agent] is an agent that is based on the jenkins/agent image above. It provides a wrapper script written in PowerShell to help specify the parameters to agent.jar.

Reviewer 3: ```suggestion
* link:https://hub.docker.com/r/jenkins/inbound-agent[jenkins/inbound-agent] is an agent that is based on the jenkins/agent image above. It provides a wrapper script written in PowerShell to help specify the parameters to agent.jar.
  Windows images are available starting from version link:https://github.com/jenkinsci/docker-inbound-agent/releases/tag/4.3-4[4.3-4]
```

> * link:https://hub.docker.com/r/jenkins/ssh-agent[jenkins/ssh-agent] is an image which has OpenSSH installed and should be used with the plugin:ssh-slaves[SSH Build Agents Plugin].

Reviewer 3: ```suggestion
* link:https://hub.docker.com/r/jenkins/ssh-agent[jenkins/ssh-agent] is an image which has OpenSSH installed and should be used with the plugin:ssh-slaves[SSH Build Agents Plugin].
  Windows images are available starting from version link:https://github.com/jenkinsci/docker-ssh-agent/releases/tag/2.1.0[2.1.0]
```

== Deeper Dive

=== jenkins/agent

This is a base image for Docker, which includes the JDK and the Jenkins agent executable (agent.jar). This executable is an instance of the Jenkins Remoting library. Multiple images are published for different versions of the JDK (8 or 11)
as well as windowsservercore and nanoserver versions. This image is mainly useful as a base image for other images (`FROM jenkins/agent:jdk8-nanoserver-1809` in your Dockerfile). When using this agent, select `Launch agent by connecting it to the master` in the `Launch method` drop-down box. You will also want to specify the `Remote root directory` as a directory you prefer, or a good default to specify is `C:\Users\jenkins\Agent`.

The agent image has a user `jenkins` which is used to run tasks on the Windows agent. The `jenkins` user is a member of the Administrators group inside the container.

When the agent is created, it will show you the information needed to pass to agent.jar to run the agent and have it connect to the master. Here is an example of that command:

```

> java -jar agent.jar -jnlpUrl https://JENKINS_URL/jenkins/computer/AGENTNAME/slave-agent.jnlp -secret SOMESECRETVALUE -workDir "C:\Users\jenkins\Agent"

Reviewer 4: Does not make much sense to me. If you were running an inbound agent (as this CLI syntax implies), why would you not just use the `inbound-agent` image? The base image is useful because you can launch it in stdio mode from all sorts of tools or scripts.

Author: I haven't used the base image very much, so I don't know much about what it is used for. If you have additional text that you can suggest, I would be more than happy to incorporate it.

Reviewer 4: https://github.com/jenkinsci/docker-agent#usage for example

Author: Ok, thanks for the clarification. I will incorporate this type of example.

* `-Secret <SECRET>` or JENKINS_SECRET - the secret as shown on the master after creating the agent
* `-Name <NAME>` or JENKINS_AGENT_NAME - the name of the agent, it should match the name you specified when creating the agent on the master
* `-Tunnel <HOST:PORT>` or JENKINS_TUNNEL - A tunnel host to route the TCP traffic through when connecting to the master
* `-WorkDir <WORKDIR>` or JENKINS_AGENT_WORKDIR - same as the -workDir parameter mentioned above in the jenkins/agent image information.
* `-WebSocket` or JENKINS_WEB_SOCKET=true - when present the connection to the master will be done via WebSocket through the Jenkins URL rather than using a separate network port.
* `-DirectConnection <HOST:PORT>` or JENKINS_DIRECT_CONNECTION - Connect directly to this TCP agent port, skipping the HTTP(S) connection parameter download.
* `-InstanceIdentity <BASE64VALUE>` or JENKINS_INSTANCE_IDENTITY - The base64 encoded InstanceIdentity byte array of the Jenkins master. When this is set, the agent skips connecting to an HTTP(S) port for connection info
* `-JavaHome <JAVA_HOME>` or JAVA_HOME  - An override for the default JAVA_HOME baked into the image (different for JDK8 vs. JDK11)

> * `-Protocols <PROTOCOLS>` or JENKINS_PROTOCOLS - Specify the link:https://github.com/jenkinsci/remoting/blob/de7818885a5bf478760ba29f5ee216291437cb16/docs/protocols.md#active-protocols[remoting protocols] to attempt when instanceIdentity is provided.

Reviewer 4: A blog post is not the place to go into this level of detail, which belongs in reference documentation somewhere—either in `jenkinsci/docker-inbound-agent` in Markdown, or on jenkins.io/doc.

[NOTE]
====

> There is currently an issue where the `-Protocols` parameter is not actually passed to agent.jar. The fix has been merged, but a release has not been made.

Reviewer 3: Add a link?

====

Example:

```
docker run jenkins/inbound-agent:windowsservercore-1809 -Url http://jenkins-server:port -WorkDir=C:/Users/jenkins/Agent -Secret <SECRET> -Name <AGENTNAME>
```

Example using environment variables:

```
docker run -e "JENKINS_URL=http://jenkins-server:port" -e "JENKINS_AGENT_NAME=AGENTNAME" jenkins/inbound-agent:windowsservercore-1809 -WorkDir=C:/Users/jenkins/Agent -Secret <SECRET> -Name <AGENTNAME>
```

> The URL, Name and Secret parameters are required, but can be specified as either command line parameters or environment variables.

Reviewer 3: Would be great to also have an example of using this image in Jenkins Pipeline.
IIRC @jetersen had an example at the Lisbon contributor summit for eval versions.

Author: I asked him via email and he no longer has access to those pipelines.

=== jenkins/ssh-agent

> As mentioned above the jenkins/ssh-agent docker image is based on SSH communication with the master, rather than the remoting TCP or WebSocket protocols. This communication is initiated by the master rather than the agent connecting to the master is done with jenkins/agent and jenkins/inbound-agent.

Reviewer 4: Again the most important use case (IMO) for `jenkins/agent` is a DiY outbound agent.

Author: Let me know if my updates are ok with regards to your ideas.

The image sets up a `jenkins` user and the OpenSSH server so that the master can connect to the agent via SSH. The image expects an SSH public key as a parameter and puts that key into the authorized_keys file for the jenkins user. The private key should be specified in the agent configuration on the master to allow the master to connect correctly.

Example:

```

> docker run jenkins/ssh-agent "<public key>"

Reviewer 4: A multiline string? Really?

Author: Not sure I understand your concern/question with this. Are you asking why a multiline string around the docker command, or if a multiline string can be passed for the public key?

Reviewer 4: The latter. Never mind. I was thinking of a _private_ key which would typically be a text file with dozens of lines. A _public_ key should indeed be single-line (albeit a long line).

Author: Ok, thanks for clarifying.

```

You can also pass the public key as an environment variable when using `docker run`.

Example:

```
docker run -e "JENKINS_AGENT_SSH_PUBKEY=<public key>" jenkins/ssh-agent
```

You will then be able to connect this agent using the plugin:ssh-slaves[SSH Build Agents Plugin] as "jenkins" with the matching private key.

== What's next?

> There is an link:github.com/jenkinsci/docker/pull/924[open PR] to create a Windows based Docker image for a Jenkins master. There hasn't been a lot of requests for this, but to make the offerings complete for Windows users, the PR was created.

Reviewer 3: ```suggestion
There is an link:github.com/jenkinsci/docker/pull/924[open pull-request] to create a Windows based Docker image for a Jenkins master. There hasn't been a lot of requests for this, but to make the offerings complete for Windows users, the PR was created.
In the case of agents, we also consider providing versions based on _Windows 10, version 1909_ so that Jenkins users could run these images on GKE clusters
(see link:https://github.com/jenkinsci/docker-agent/issues/134[this issue]).

For plans unrelated to Windows, please see the link:/blog/2020/05/06/docker-agent-image-renaming/[Docker images for agents: New names and What's next] blogpost.
```

Follow-up from the platform SIG meeting discussion about GKE.
